src/modm/platform/can/atsam-fdcan/can_impl.hpp

```diff
@@ -0,0 +1,637 @@
+
+	configureMode(startupMode);
+}
+
+template<uint8_t id, fdcan::MessageRamConfig mrc>
+void
+McanDriver<id, mrc>::setMode(Mode mode)
+{
+	EnterInitMode init;
+	Regs()->MCAN_CCCR |= MCAN_CCCR_CCE;
+	configureMode(mode);
+}
+
+template<uint8_t id, fdcan::MessageRamConfig mrc>
+void
+McanDriver<id, mrc>::configureMode(Mode mode)
+{
+	// Reset all mode register bits
+	Regs()->MCAN_TEST = 0;
+	Regs()->MCAN_CCCR &= ~(MCAN_CCCR_ASM | MCAN_CCCR_MON | MCAN_CCCR_CSR | MCAN_CCCR_TEST);
+
+	// set mode
+	switch(mode) {
+		case Mode::Normal:
+			break;
+		case Mode::Restricted:
+			Regs()->MCAN_CCCR |= MCAN_CCCR_ASM;
+			break;
+		case Mode::Monitoring:
+			Regs()->MCAN_CCCR |= MCAN_CCCR_MON;
+			break;
+		case Mode::Sleep:
+			Regs()->MCAN_CCCR |= MCAN_CCCR_CSR;
+			break;
+		case Mode::TestExternalLoopback:
+			Regs()->MCAN_CCCR |= MCAN_CCCR_TEST;
+			Regs()->MCAN_TEST = MCAN_TEST_LBCK;
+			break;
+		case Mode::TestInternalLoopback:
+			Regs()->MCAN_CCCR |= MCAN_CCCR_TEST | MCAN_CCCR_MON;
+			Regs()->MCAN_TEST = MCAN_TEST_LBCK;
+			break;
+	}
+}
+
+template<uint8_t id, fdcan::MessageRamConfig mrc>
+void
+McanDriver<id, mrc>::setAutomaticRetransmission(bool retransmission)
+{
+	if (retransmission) {
+		// Enable retransmission
+		Regs()->MCAN_CCCR = (Regs()->MCAN_CCCR & ~MCAN_CCCR_DAR);
+	} else {
+		// Disable retransmission
+		Regs()->MCAN_CCCR = (Regs()->MCAN_CCCR | MCAN_CCCR_DAR);
+	}
+}
+
+template<uint8_t id, fdcan::MessageRamConfig mrc>
+bool
+McanDriver<id, mrc>::isMessageAvailable()
+{
+	return rxFifo0HasMessage() || rxFifo1HasMessage();
+}
+
+template<uint8_t id, fdcan::MessageRamConfig mrc>
+bool
+McanDriver<id, mrc>::getMessage(can::Message& message, uint8_t *filter_id, uint16_t *timestamp)
+{
+	if (rxFifo0HasMessage()) {
+		readMsg(message, 0, filter_id, timestamp);
+		return true;
+	} else if (rxFifo1HasMessage()) {
+		readMsg(message, 1, filter_id, timestamp);
+		return true;
+	}
+	return false;
+}
+
+template<uint8_t id, fdcan::MessageRamConfig mrc>
+bool
+McanDriver<id, mrc>::isReadyToSend()
+{
+	return !isHardwareTxQueueFull();
+}
+
+template<uint8_t id, fdcan::MessageRamConfig mrc>
+bool
+McanDriver<id, mrc>::sendMessage(const can::Message& message)
+{
+	return sendMsg(message);
+}
+
+template<uint8_t id, fdcan::MessageRamConfig mrc>
+McanDriver<id, mrc>::BusState
+McanDriver<id, mrc>::getBusState()
+{
+	if (Regs()->MCAN_PSR & MCAN_PSR_BO) {
+		return BusState::Off;
+	}
+	else if (Regs()->MCAN_PSR & MCAN_PSR_EP) {
+		return BusState::ErrorPassive;
+	}
+	else if (Regs()->MCAN_PSR & MCAN_PSR_EW) {
+		return BusState::ErrorWarning;
+	}
+	else {
+		return BusState::Connected;
+	}
+}
+
+template<uint8_t id, fdcan::MessageRamConfig mrc>
+bool
+McanDriver<id, mrc>::setStandardFilter(
+	uint8_t standardIndex, FilterConfig config,
+	modm::can::StandardIdentifier id_,
+	modm::can::StandardMask mask)
+{
+	if (standardIndex >= MessageRam::StandardFilterCount) {
+		return false;
+	}
+
+	// TODO: Is INIT mode really necessary to add/modify filters?
```

Comment:
I can't find any hint the in datasheet INIT mode would be necessary, and accessing the filter (located in message RAM in standard SRAM) should not cause any issues.
In the worst case some messages that should not get accepted will get accepted if the MCAN peripheral reads the filter while it is being written.

Reply:
Could you have some weird behavior when you receive a message while updating filters?

Reply:
The standard filter is safe because the whole config is a single 32 bit word, but the extended filter consists of two 32 bit words.
If we write the word containing the enable bit(s) to zero first, then update the other filter config word and finally write the word with the enable (==mode selection) bits, we should be safe.

Reply:
I fixed it in 65ec533a4.